src/container/Container.js

```diff
@@ -2,228 +2,347 @@ import { _, Promise } from '../globals';
 import ExtensionApi from './ExtensionApi';
 import { ServiceError } from './errors';
 
-function countOccurrences (array, item) {
-    return array.filter(_.isEqual.bind(_, item)).length;
+export default function Container (extensions, config) {
```

Comment:
By "new style", do refer to it's not an ES6 class anymore?

If yes, then I only changed it, because a simple object is easy to compose, extend and copy. I prefer the ES6 class syntax better as well, however as it is creating non-enumerable properties on the prototype from it's methods we cannot use it in the `ExtensionApi` module as we do right now. What we could do there though (`ExtensionApi`) that instead of extending we instantiate a new `Container` based on the previous one and that should work. (Although I am not sure yet that we actually need to create a new object there tbh)

Regarding the private functions, do you refer to the actual private keywords (e.g. in TypeScript) or just module scope functions which are not public?
Regarding the helper functions, I actually want to skip using `call` and assigning a context as much as possible, and just keeping the functions depending only on their parameters rather than on the context. I think that's less error-prone and easier to think about.